### PR TITLE
feat: split vacations to separate collection

### PR DESCRIPTION
**Pull request by Shelex**

### CODE REVIEW

File: client/graph/factory/transform.go [hunk 1]
```diff
@@ -21,6 +21,10 @@ func EmployeeFromProtoToApi(e *proto.Employee) *model.Employee {
 	for i := range apiDocs {
 		apiDocs[i] = &e.Documents[i]
 	}
+	vacations := make([]*string, len(e.Vacations))
+	for i := range vacations {
+		vacations[i] = &e.Vacations[i]
```
Comment (the4thamigo-uk) [inline client/graph/factory/transform.go]: Small thing: you could use `for i,v := range vacations` and use the `v` on line 26

File: domain/entities/entities.go [hunk 2]
```diff
@@ -6,22 +6,23 @@ import (
...
-	Vacations           []Vacation `json:"vacations"`
-	Documents           []string   `json:"documents"`
+	ID                  string   `json:"id" bson:"_id"`
+	BadgeNumber         int32    `json:"badgeNumber" bson:"badgeNumber"`
+	FirstName           string   `json:"firstName" bson:"firstName"`
+	LastName            string   `json:"lastName" bson:"lastName"`
+	CountryCode         string   `json:"countryCode" bson:"countryCode"`
+	VacationAccrualRate float32  `json:"vacationAccrualRate" bson:"vacationAccrualRate"`
+	VacationAccrued     float32  `json:"vacationAccrued" bson:"vacationAccrued"`
+	Vacations           []string `json:"vacations"`
```
Comment (the4thamigo-uk) [inline domain/entities/entities.go]: no bson struct tags here?

File: domain/main.go [hunk 3]
```diff
@@ -211,10 +211,10 @@ func (e *employeeService) UpdateEmployee(ctx context.Context, req *proto.Employe
 
 func (e *employeeService) AddVacation(ctx context.Context, req *proto.VacationRequest) (*proto.Vacation, error) {
 	start := time.Unix(req.StartDate, 0)
-	tomorrow := time.Now().Add(24 * time.Hour)
+	hours24 := time.Now().Add(24 * time.Hour)
 
-	afterTomorrow := start.After(tomorrow)
-	if !afterTomorrow {
+	after24Hours := start.After(hours24)
+	if !after24Hours {
```
Comment (the4thamigo-uk) [inline domain/main.go]: you can put both lines in a single if statement, as you dont need `after24Hours` later in the function

File: domain/main.go [hunk 4]
```diff
@@ -226,3 +226,16 @@ func (e *employeeService) AddVacation(ctx context.Context, req *proto.VacationRe
...
 }
+
+func (e *employeeService) Vacations(req *proto.GetAllRequest, stream proto.EmployeeService_VacationsServer) error {
+	vacations, err := e.repository.Vacations()
+	if err != nil {
+		return err
+	}
+	for _, vacation := range vacations {
+		if err := stream.Send(entities.VacationFromStorageToProto(vacation)); err != nil {
+			return err
```
Comment (the4thamigo-uk) [inline domain/main.go]: Think about wrapping errors where it makes sense, to improve debugging. https://blog.golang.org/go1.13-errors

File: storage/inmem.go [hunk 5]
```diff
@@ -160,9 +163,18 @@ func (i *InMem) AddVacation(ID string, userID string, startDate int64, durationH
...
+	v.UserID = userID
 	v.StartDate = startDate
 	v.DurationHours = durationHours
-	employee.Vacations = append(employee.Vacations, v)
+	employee.Vacations = append(employee.Vacations, v.ID)
 	i.employees[userID] = employee
 	return v, nil
 }
+
+func (i *InMem) Vacations() ([]entities.Vacation, error) {
```
Comment (the4thamigo-uk) [inline storage/inmem.go:175]: Is this just the `copy()` built-in function?

File: storage/mongo.go [hunk 6]
```diff
@@ -157,23 +160,36 @@ func (r *Repository) AddDocument(userID string, ID string) error {
 func (r *Repository) AddVacation(ID string, userID string, startDate int64, durationHours float32) (entities.Vacation, error) {
 	session := r.session.Copy()
 	defer session.Close()
```
Comment (the4thamigo-uk) [inline storage/mongo.go:156]: does `session.Close()` return an error? If so we probably ought not to ignore it? Here is some nice pattern for this https://www.joeshaw.org/dont-defer-close-on-writable-files/

File: storage/mongo.go [hunk 7]
```diff
@@ -157,23 +160,36 @@ func (r *Repository) AddDocument(userID string, ID string) error {
...
-	update[vacationIndex] = v
+	update[vacationIndex] = vacation.ID
+
+	if err := session.DB(r.name).C(r.vacations).Insert(vacation); err != nil {
+		return vacation, errors.Wrap(err, "failed to create vacation")
+	}
 
 	if err := session.DB(r.name).C(r.employees).UpdateId(userID, bson.M{"$set": update}); err != nil {
-		return v, err
+		return vacation, errors.Wrap(err, "failed to update employee vacations")
```
Comment (the4thamigo-uk) [inline storage/mongo.go:177]: you can use `fmt.Errorf` with `%w` nowadays. Its a bit ugly I think, but it is now the idiomatic way I believe.

File: storage/inmem.go (hunk 5 above)
Comment (Shelex) [reply storage/inmem.go:175]: Basically it is writing of map values to slice.
The map is used to ease operations with vacations by id.

File: storage/mongo.go (hunk 6 above)
Comment (Shelex) [reply storage/mongo.go:156]: `mgo.session.Close()` does not return error, but link is helpful to understand that such errors should be handled in other cases :thumbsup:

Comment (Shelex) [reply storage/mongo.go:156]: As example, refactored grpc.ClientConn:
``` defer conn.Close()```
to
```
defer func() {
		if err := conn.Close(); err != nil {
			log.Fatalf("failed to close connection: %s", err)
		}
	}()
```

File: storage/mongo.go (hunk 7 above)
Comment (Shelex) [reply storage/mongo.go:177]: And anyway allows to check wrapped errors which is nice.

File: storage/inmem.go (hunk 5 above)
Comment (the4thamigo-uk) [reply storage/inmem.go:175]: oh yes sorry I misread that one.

**Pull request merged**